src/charm.py

```diff
@@ -308,6 +311,38 @@ def _validate(self):
         if not self._state.schema_ready:
             raise ValueError("admin:temporal relation: schema is not ready")
 
+    def _open_service_ports(self):
+        """Open the respective ports based on Temporal service."""
```

Comment:
Alternatively, and more concisely:
```python
SERVICE_PORTS = {
    "matching": {
        "grpc": 7235,
        "http": 6935, 
    },
    "worker": {
        "grpc": 7239,
        "http": 6939, 
    },
    ...
}

open_port = functools.partial(self.model.unit.open_port, protocol="tcp")
close_port = functools.partial(self.model.unit.close_port, protocol="tcp")

for service, ports in SERVICE_PORTS.items():
    if service in services:
        open_port(service["grpc"])
        open_port(service["http"])
    else:
        close_port(service["grpc"])
        close_port(service["http"])

```

Reply:
Thanks for the advice! Will include it in the next PR.